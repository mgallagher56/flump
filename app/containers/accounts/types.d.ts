import type { Database } from 'db_types';

export type Account = Partial<Database['public']['Tables']['accounts']['Row']>;
export type AccountDetail = Partial<Database['public']['Tables']['account_details']['Row']>;
