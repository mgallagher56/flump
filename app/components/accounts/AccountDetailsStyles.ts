import { css } from 'styled-system/css';

export const accountDetailContainerStyles = css({
  overflow: 'auto'
});

export const accountDetailDisplayStyles = css({
  display: 'flex',
  flexDirection: 'row',
  gap: 5
});

export const flexColumnStyles = css({
  display: 'flex',
  flexDirection: 'column',
  gap: 5
});
