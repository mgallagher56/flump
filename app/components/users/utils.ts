import type { SupabaseClient } from '@supabase/supabase-js';
import type { TFunction } from 'i18next';

export enum SignUpActionEnum {
  LOGIN = 'logIn',
  SIGNUP = 'signUp'
}

export const getSignUpButtonText = (t: TFunction, isShowLogin: boolean, isMagicLink: boolean) => {
  if (isMagicLink) return t('sendMagicLink');
  if (isShowLogin) return t(SignUpActionEnum.LOGIN);
  return t(SignUpActionEnum.SIGNUP);
};

export const getInfoMessage = (t: TFunction, isUserRegistered: boolean, isMagicLink: boolean) => {
  if (isMagicLink) return t('magicLinkSent');
  if (isUserRegistered) return t('clickAgainToLogIn');
  return null;
}

export const getSubmitUserAuthAction = async ({
  supabaseClient,
  isShowLogin,
  isMagicLink,
  formInput
}: {
  supabaseClient: SupabaseClient;
  isShowLogin: boolean;
  isMagicLink: boolean;
  formInput: { email: string; password: string };
}) => {
  if (isMagicLink)
    return await supabaseClient.auth.signInWithOtp({
      email: formInput.email,
      options: {
        emailRedirectTo: window.location.origin + '/auth/callback'
      }
    });
  if (isShowLogin) return await supabaseClient.auth.signInWithPassword(formInput);
  return await supabaseClient.auth.signUp(formInput);
};
