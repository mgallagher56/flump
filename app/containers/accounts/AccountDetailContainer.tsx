import type { FC } from 'react';

import { useLoaderData } from '@remix-run/react';
import FLPHeading from '~/components/core/typography/FLPHeading';
import type { loader } from '~/routes/app.accounts.$account';

const AccountDetailContainer: FC = () => {
  const { account } = useLoaderData<typeof loader>();
  console.log({ account });
  return (
    <>
      <FLPHeading as="h2" size="sm">{account?.type}</FLPHeading>
      <FLPHeading as="h1" size="xl" >{account?.name}</FLPHeading>
    </>
  );
};

export default AccountDetailContainer;
