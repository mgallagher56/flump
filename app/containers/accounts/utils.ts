import type { Account } from './types';

export enum AccountTypeEnum {
  CURRENT = 'Current',
  SAVING = 'Saving',
  MORTGAGE = 'Mortgage',
  LOAN = 'Loan',
  OWED = 'Owed'
}

export const isAccountTypeValid = (accountType: AccountTypeEnum, accounts: Account[]) =>
  !!accounts?.map(({ type }) => type).some((type) => type === accountType);
