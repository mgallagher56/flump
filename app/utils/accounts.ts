export const monthYearSort = (
  a: {
    month?: number;
    year?: number;
  },
  b: {
    month?: number;
    year?: number;
  }
) => {
  if (a.year === b.year) {
    return a.month - b.month;
  }
  return a.year - b.year;
};
