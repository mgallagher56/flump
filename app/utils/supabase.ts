import { createBrowserClient, createServerClient } from '@supabase/auth-helpers-remix';

import type { Database } from 'db_types';

const isServer = typeof window === 'undefined';

export const getSupaBaseUrl = (isServer: boolean): string =>
  isServer ? process.env.SUPABASE_URL : window?.env?.SUPABASE_URL;
export const getSupabaseAnonKey = (isServer: boolean): string =>
  isServer ? process.env.SUPABASE_ANON_KEY : window?.env?.SUPABASE_ANON_KEY;

const supabase = createBrowserClient<Database>(getSupaBaseUrl(isServer), getSupabaseAnonKey(isServer));

export const createSupaBaseServerClient = ({ request, response }: { request: Request; response: Response }) =>
  createServerClient<Database>(getSupaBaseUrl(isServer), getSupabaseAnonKey(isServer), { request, response });

export default supabase;
