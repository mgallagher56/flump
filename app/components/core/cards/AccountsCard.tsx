import { type FC, useCallback } from 'react';

import type { CardProps } from '@chakra-ui/react';
import { useLoaderData, useNavigate, useRevalidator } from '@remix-run/react';
import { useTranslation } from 'react-i18next';
import type { AccountTypeEnum } from '~/containers/accounts/utils';
import AddEditAccountsDialogBtn from '~/containers/dialogs/addEditAccountsDialog.tsx/AddEditAccountsDialog';
import type { loader } from '~/root';
import supabase from '~/utils/supabase';

import FLPButton from '../buttons/FLPButton';
import FLPButtonGroup from '../buttons/FLPButtonGroup';
import FLPBox from '../structure/FLPBox';
import FLPText from '../typography/FLPText';
import FLPCard from './FLPCard';

interface AccountsCardProp extends Omit<CardProps, 'title'> {
  accountId: string;
  name: string;
  type: AccountTypeEnum;
}

const AccountsCard: FC<AccountsCardProp> = ({ accountId, name, type }) => {
  const { t } = useTranslation();
  const navigate = useNavigate();
  const { user } = useLoaderData<typeof loader>();
  const { revalidate } = useRevalidator();

  const handleRemoveAccount = useCallback(
    async (e) => {
      e.stopPropagation();
      await supabase.from('accounts').delete().eq('user_id', user.id).eq('id', accountId);
      revalidate();
    },
    [accountId, revalidate, user.id]
  );

  const onCardClick = useCallback(() => navigate(`/app/accounts/${accountId}`), [accountId, navigate]);

  return (
    <FLPCard
      id={accountId}
      title={name}
      padding={5}
      marginY={5}
      width="250px"
      height="300px"
      display="flex"
      direction="column"
      justifyContent="space-between"
      gap={2}
      onClick={onCardClick}
    >
      <FLPBox display="flex" gap="3" flexDirection="column">
        <FLPText fontWeight="bold">{name}</FLPText>
        <FLPText>{`${type} ${t('account')}`}</FLPText>
      </FLPBox>
      <FLPButtonGroup justifyContent="flex-end" zIndex="10">
        <AddEditAccountsDialogBtn accountId={accountId} isEditAccount={true} />
        <FLPButton colorScheme="red" onClick={handleRemoveAccount}>
          {t('remove')}
        </FLPButton>
      </FLPButtonGroup>
    </FLPCard>
  );
};

export default AccountsCard;
