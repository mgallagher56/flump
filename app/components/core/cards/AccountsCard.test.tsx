import { fireEvent, render } from '@testing-library/react';
import { vi } from 'vitest';
import { AccountTypeEnum } from '~/containers/accounts/utils';

import mockUser from '__mocks__/user';

import AccountsCard from './AccountsCard';

const mocks = vi.hoisted(() => ({
  mockUseLoaderData: vi.fn(),
  mockUseRevalidator: vi.fn(() => ({ revalidate: vi.fn() })),
  mockUseNavigate: () => vi.fn(),
  mockFrom: vi.fn(() => ({
    delete: () => ({
      eq: () => ({
        eq: () => ({})
      })
    })
  }))
}));

vi.mock('@remix-run/react', async () => {
  const actual: Record<string, unknown> = await vi.importActual('@remix-run/react');
  return {
    ...actual,
    useLoaderData: mocks.mockUseLoaderData,
    useRevalidator: mocks.mockUseRevalidator,
    useNavigate: mocks.mockUseNavigate
  };
});

vi.mock('app/utils/supabase', () => ({
  default: {
    from: mocks.mockFrom
  }
}));

describe('<AccountsCard />', () => {
  mocks.mockUseLoaderData.mockReturnValue({ user: mockUser });

  test('it renders an AccountsCard component with title as expected', () => {
    const { baseElement } = render(
      <AccountsCard accountId={'123456'} name="My curent account" type={AccountTypeEnum.CURRENT} />
    );
    expect(baseElement).toMatchSnapshot();
  });
  test('it calls supabase.delete when delete button is clicked', () => {
    const { getByText } = render(
      <AccountsCard accountId={'123456'} name="My curent account" type={AccountTypeEnum.CURRENT} />
    );
    const deleteButton = getByText('remove');
    fireEvent.click(deleteButton);
    expect(mocks.mockFrom).toBeCalledWith('accounts');
  });
});
