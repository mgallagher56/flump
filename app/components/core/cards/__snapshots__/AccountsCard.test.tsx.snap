// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<AccountsCard /> > it renders an AccountsCard component with title as expected 1`] = `
<body>
  <div>
    <div
      class="chakra-card pos_relative transition_all_0.3s_ease-in-out border_1px_solid_token(colors.gray.500)! [&:hover]:shadow_2px_5px_8px_0px_token(colors.gray.500)! [&:hover]:transform_translate3d(-2px,_-5px,_-5px) [&:hover]:cursor_pointer css-1r1arwj"
      id="123456"
      title="My curent account"
    >
      <div
        class="css-5ivlx4"
      >
        <p
          class="chakra-text css-hmjgxl"
        >
          My curent account
        </p>
        <p
          class="chakra-text css-km77ca"
        >
          Current account
        </p>
      </div>
      <div
        class="chakra-button__group css-vhnxbm"
        data-orientation="horizontal"
        role="group"
      >
        <button
          class="chakra-button css-1fysbuj"
          type="button"
        >
          edit
        </button>
        <button
          class="chakra-button css-1fysbuj"
          type="button"
        >
          remove
        </button>
      </div>
    </div>
  </div>
</body>
`;
